virtual_rainforest/core/model.py

```diff
@@ -0,0 +1,116 @@
+
+    Describes the common functions and attributes that all `vr` models should have. This
+    includes functions to setup, spin up and solve the specific model, as well as a
+    function to cleanup redundant model data. At this level these functions are not
+    define and are mere placeholders to be overwritten (where appropriate) by the
+    inheriting classes.
+
+    Args:
+        start_time: Point in time that the model simulation should be started.
+        end_time: Time that the model simulation should end
+        update_interval: Time to wait between updates of the model state.
+
+    Attributes:
+        name: Names the model that is described
+    """
+
+    name = "base"
+    # TODO - Once higher level timing function is written use it to set this
+    last_update = datetime64("2000-01-01")
+
+    def __init__(self, *args: Any, **kwargs: Any):
+        # Save args and kwargs, so that they can be found by the __repr__
+        self._args = args
+        self._kwargs = kwargs
+
+        self.update_interval: timedelta64 = args[0]
+
+    @abstractmethod
+    def setup(self) -> None:
+        """Function to use input data to set up the model."""
+
+    @abstractmethod
+    def spinup(self) -> None:
+        """Function to spin up the model."""
+
+    @abstractmethod
+    def solve(self) -> None:
+        """Function to solve the model."""
+
+    @abstractmethod
+    def cleanup(self) -> None:
+        """Function to delete objects within the class that are no longer needed."""
+
+    @classmethod
+    @abstractmethod
+    def factory(cls, config: dict[str, Any]) -> Any:
+        """Factory function to unpack config and initialise a model instance."""
+
+    @classmethod
+    def __init_subclass__(cls, model_name: str):
+        """Method that adds new model classes to the model registry."""
+
+        # Add the grid type to the registry
+        if model_name in MODEL_REGISTRY:
+            LOGGER.warning(
+                "Model with name %s already exists and is being replaced", model_name
+            )
+        MODEL_REGISTRY[model_name] = cls
+
+    def should_update(self, current_time: datetime64) -> bool:
+        """Determines whether a model should be updated for a specific time step."""
+
+        if current_time > self.last_update + self.update_interval:
+            self.last_update = current_time
+            return True
+        return False
+
+    def __repr__(self) -> str:
+        """Represent a Model as a string."""
+
+        # Add all args to the function signature
+        func_sig = ""
+        for idx, arg in enumerate(self._args):
+            if idx == 0:
+                func_sig += f"{arg},"
+            else:
+                func_sig += f" {arg},"
+
+        if len(self._kwargs) > 0:
+            # TODO - Work out how to print kwargs nicely, (once we have some)
+            func_sig += f" {self._kwargs}"
+        else:
+            # Remove final comma
+            func_sig = func_sig[:-1]
```

Comment:
A better version, following my last comment above:
```suggestion
        func_sig = ", ".join([f"{k} = {getattr(self, k)}" for k in self._repr])
```